pkg/quota/image/imagestreamimport_evaluator.go

```diff
@@ -0,0 +1,220 @@
+		MatchedResourceNames:       computeResources,
+		MatchesScopeFunc:           matchesScopeFunc,
+		UsageFunc:                  makeImageStreamImportAdmissionUsageFunc(osClient, maximumTagsPerRepo),
+		ConstraintsFunc:            imageStreamImportConstraintsFunc,
+	}
+}
+
+// imageStreamImportConstraintsFunc checks that given object is an image stream import.
+func imageStreamImportConstraintsFunc(required []kapi.ResourceName, object runtime.Object) error {
+	if _, ok := object.(*imageapi.ImageStreamImport); !ok {
+		return fmt.Errorf("unexpected input object %v", object)
+	}
+	return nil
+}
+
+// imageStreamImportUsageComputer computes resource usage of image stream import objects.
+type imageStreamImportUsageComputer struct {
+	GenericImageStreamUsageComputer
+	// we assume that import of a repository will result in a maximal number of tags possible specified by
+	// this value
+	maximumTagsPerRepo int
+	processedSpecRefs  sets.String
+	processedRefs      sets.String
+}
+
+// makeImageStreamImportAdmissionUsageFunc retuns a function for computing a usage of an image stream import.
+func makeImageStreamImportAdmissionUsageFunc(osClient osclient.Interface, maximumTagsPerRepo int) generic.UsageFunc {
+	f := makeImageStreamImportUsageComputerFactory(osClient, maximumTagsPerRepo)
+	return func(object runtime.Object) kapi.ResourceList {
+		return f().Usage(object)
+	}
+}
+
+// makeImageStreamImportUsageComputerFactory returns an object used during computation of image quota across all
+// repositories in a namespace.
+func makeImageStreamImportUsageComputerFactory(osClient osclient.Interface, maximumTagsPerRepo int) quotautil.UsageComputerFactory {
+	return func() quotautil.UsageComputer {
+		return &imageStreamImportUsageComputer{
+			GenericImageStreamUsageComputer: *NewGenericImageStreamUsageComputer(osClient),
+			maximumTagsPerRepo:              maximumTagsPerRepo,
+			processedSpecRefs:               sets.NewString(),
+			processedRefs:                   sets.NewString(),
+		}
+	}
+}
+
+// makeImageStreamImportAdmissionUsageFunc returns a function that computes a resource usage of image stream
+// import objects. It is being used solely in the context of admission check for CREATE operation on
+// ImageStreamImport kind.
+func (c *imageStreamImportUsageComputer) Usage(object runtime.Object) kapi.ResourceList {
+	isi, ok := object.(*imageapi.ImageStreamImport)
+	if !ok {
+		return kapi.ResourceList{}
+	}
+
+	usage := map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *resource.NewQuantity(0, resource.DecimalSI),
+		imageapi.ResourceImageStreamImages: *resource.NewQuantity(0, resource.DecimalSI),
+	}
+
+	if !isi.Spec.Import || (len(isi.Spec.Images) == 0 && isi.Spec.Repository == nil) {
+		return usage
+	}
+
+	// collect all the image references already tagged in a project
+	iss, err := c.osClient.ImageStreams(isi.Namespace).List(kapi.ListOptions{})
```

Comment:
Can you add a TODO here that this could be removed in one of two cases:

1. We have a shared cache on the backend for image streams and/or unique referenced images per namespace that would allow fast checks
2. Instead of having this extra admission controller for images, we used a wrapper admission controller on the storage object passed into ImageStreamTags, ImageStreamImports, and ImageStreamMappings that only knew about ImageStreams and could do the accurate check.

We had discussed the second option originally - given the complexity of this, post 3.2 I'd like to refactor down to that simpler path and invoke a specific "admission storage wrapper" as discussed with @csrwng and @ncdc several times (I believe we'd done a real prototype).  That would mean that we only need to calculate ImageStreams and all of the other evaluators can be removed.